lib/fluent/winsvc.rb

```diff
@@ -57,8 +59,16 @@ def initialize(service_name)
     
     def service_main
       @pid = service_main_start(@service_name)
-      while running?
-        sleep 10
+      begin
+        loop do
+          sleep 5
+          break unless running?
+          pid, status = Process.waitpid2(@pid, Process::WNOHANG)
+          raise Errno::ECHILD if pid || status
+        end
+      rescue Errno::ECHILD
+        set_event(@@hStopEvent)
```

Comment:
```suggestion
      rescue Errno::ECHILD
        @pid = 0
        set_event(@@hStopEvent)
```

`service_stop` callback is called after `hStopEvent`.
We should set `@pid` to `0` to skip the callback because the callback will wait for the supervisor process to init, meaninglessly, for a while.

https://github.com/fluent/fluentd/blob/c3ebd6687931298a3d6faf7d0e8383f31d81357c/lib/fluent/winsvc.rb#L65-L72

In addition, it would be better to change this callback as follows.

```rb
 def service_stop 
   return if @pid <= 0 
   wait_supervisor_finished 
 end 
```

Reply:
Done. More simplified.

```ruby
def
  wait_supervisor_finished if @pid > 0
end
```

Reply:
Thanks! :heart: